vagrant: dev-env based on latest kernel

nyrahul opened the pull request:
A dev-env based on the latest kernel is required to test/dev corresponding latest features from the kernel.

Fixes: #153

nyrahul committed:
vagrant: dev-env based on latest kernel

A dev-env based on latest kernel is required for features depending on
the latest kernel (for e.g, KRSI, Event-Auditor).

This patch uses Cilium's net-next ubuntu image.

Fixes: #153

Signed-off-by: Rahul Jadhav <r@accuknox.com>

geyslan commented on contribution/vagrant/Vagrantfile:
```diff
@@ -1,18 +1,27 @@
+end
+
 Vagrant.configure("2") do |config|
   config.vm.define "KubeArmor-Dev" do |cfg|
```
I think that `config.vm.define` should be the same as `VM_NAME`, so the user can up and connect to both VMs interchangeably.

```diff
 Vagrant.configure("2") do |config|
-  config.vm.define "KubeArmor-Dev" do |cfg|
+  config.vm.define VM_NAME do |cfg|
```

If it is set statically, the right `VM_IMG` won't be up in despite of `NETNEXT` when other is already created.

nyrahul replied:
Handled, thanks

geyslan reviewed (commented):
I made a single suggestion about `VM_NAME`.

nyrahul committed:
Changed the image to use ubuntu/bionic64 with latest kernel.

geyslan commented on contribution/vagrant/Vagrantfile:
```diff
@@ -35,4 +43,9 @@ Vagrant.configure("2") do |config|
+
+  # Reboot to use new kernel
+  if ENV['NETNEXT'] == "1" then
+	  config.vm.provision :reload
```
```bash
vm:
* The 'reload' provisioner could not be found.
```

It's necessary to install the plugin `reload`.

```diff
@@ -9,6 +9,7 @@ else
 end
 
 Vagrant.configure("2") do |config|
+  config.vagrant.plugins = "vagrant-reload"
   config.vm.define VM_NAME do |cfg|
     cfg.vm.box = VM_IMG
     cfg.vm.provider "virtualbox" do |vb|
@@ -46,6 +47,6 @@ Vagrant.configure("2") do |config|
 
   # Reboot to use new kernel
   if ENV['NETNEXT'] == "1" then
-     config.vm.provision :reload
+    config.vm.provision :reload
   end
 end
```

But reloading is causing ownership loss.

```bash
vagrant@kubearmor-dev-next:~$ ls -l ~/
total 8
drwxr-xr-x 1 root    root    4096 Jun 19 14:41 KubeArmor
drwxr-xr-x 4 vagrant vagrant 4096 Jun 21 23:23 go
```

And it seems tricky to solve: https://stackoverflow.com/questions/23894739/synced-folders-lost-when-rebooting-a-vagrant-machine-using-the-ansible-provision

nyrahul replied:
Yes, reloading is necessary because we are updating the kernel and we need to switch to the new one.

Reloading (even if you do vagrant halt/up) is causing ownership loss if you have an old VirtualBox-guest. If you try the `NETNEXT=1 vagrant plugin install vagrant-vbguest` this will update the VirtualBox-guest image and things work after that. The other option is for the user to use sudo when running kubearmor.

There is no other easy way to solve this problem. Using NFS to mount the synced_folder would be another option but it requires some other configuration.

nyrahul replied:
I have updated the documentation to explain these points.

geyslan replied:
This looks great. One other way would be to declare the needed plugins inside the Vagrantfile.

```diff
@@ -9,6 +9,7 @@ else
 end
 
 Vagrant.configure("2") do |config|
+  config.vagrant.plugins = ["vagrant-vbguest", "vagrant-reload"]
   config.vm.define VM_NAME do |cfg|
```

It will ask one to install plugins locally and to run the requested command again.

```bash
❯ NETNEXT=1 vagrant up
Vagrant has detected project local plugins configured for this
project which are not installed.

  vagrant-vbguest
Install local plugins (Y/N) [N]: y
Installing the 'vagrant-vbguest' plugin. This can take a few minutes...
Fetching micromachine-3.0.0.gem
Fetching vagrant-vbguest-0.30.0.gem
Installed the plugin 'vagrant-vbguest (0.30.0)'!


Vagrant has completed installing local plugins for the current Vagrant
project directory. Please run the requested command again.
```

nyrahul replied:
This is even better. Will try to handle this as part of another PR. Thanks

nyrahul committed:
Documentation: Vagrant plugins for NETNEXT

Improved the documentation to mention about vagrant plugin deps for
NETNEXT image.

nyrahul committed:
Merge branch 'master' of github.com:accuknox/KubeArmor
